File: 0200-number-of-islands/0200-number-of-islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        ROWS = len(grid)
        COLS = len(grid[0])
        visited = set()
        island = 0

        def bfs(r, c) :
            q = deque()
            visited.add((r, c))
            q.append((r, c))

            direction = [[1, 0], [-1, 0], [0, 1], [0, -1]]
            while q:
                row, col = q.popleft()
                for dr, dc in direction:
                    nr = row + dr
                    nc = col + dc
                    if (nr in range(ROWS) and
                        nc in range(COLS) and
                        grid[nr][nc] == "1" and
                        (nr, nc) not in visited) :
                        q.append((nr, nc))
                        visited.add((nr, nc))


        for r in range(ROWS):
            for c in range(COLS):
                if (grid[r][c] == "1" and  (r, c) not in visited) :
                    bfs(r, c)
                    island += 1
        
        return island

File: 0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    vector<pair<int, int>> dirs = {
        {-1, 0}, {1, 0}, {0, -1}, {0, 1}
    };

    int numIslands(vector<vector<char>>& grid) {
        auto inside = [&](int x, int y) {
            return x >= 0 && x < grid.size() && y >= 0 && y < grid[0].size();
        };
        auto dfs = [&](auto &self, int x, int y) -> void {
            for(auto [dx, dy] : dirs) {
                int ddx = x + dx, ddy = y + dy;
                if(!inside(ddx, ddy)) continue;
                if(grid[ddx][ddy] == '0') continue;
                grid[ddx][ddy] = '0';
                self(self, ddx, ddy);
            }
            return;
        };
        int cnt = 0;
        for(int i = 0; i < grid.size(); i++) {
            for(int j = 0; j < grid[0].size(); j++) {
                if(grid[i][j] == '1') {
                    dfs(dfs, i, j);
                    cnt++;
                }
            }
        }
        return cnt;
    }
};